chamge/amqp-hub-backend.c

```diff
@@ -887,6 +924,141 @@ chamge_amqp_hub_backend_deactivate (ChamgeHubBackend * hub_backend)
+  amqp_exchange_name =
+      g_settings_get_string (self->settings, "enroll-exchange-name");
+
+  g_debug ("[config] channel : %d, enroll-exchange-name : %s",
+      amqp_channel, amqp_exchange_name);
+
+  amqp_conn = amqp_new_connection ();
+  amqp_socket = amqp_tcp_socket_new (amqp_conn);
+  g_assert_nonnull (amqp_socket);
+
+  if (amqp_socket_open (amqp_socket,
+          connection_info.host, connection_info.port)) {
+    g_set_error_literal (error, CHAMGE_BACKEND_ERROR,
+        CHAMGE_BACKEND_ERROR_OPERATION_FAILURE, "socket open failure");
+    goto out;
+  }
+
+  /* TODO: The authentication method should be EXTERNAL
+   * if we don't want to share user/passwd
+   */
+  amqp_r = amqp_login (amqp_conn, connection_info.vhost, 0, 131072, 0,
+      AMQP_SASL_METHOD_PLAIN, connection_info.user, connection_info.password);
+  if (amqp_r.reply_type != AMQP_RESPONSE_NORMAL) {
+    g_set_error (error, CHAMGE_BACKEND_ERROR,
+        CHAMGE_BACKEND_ERROR_OPERATION_FAILURE, "login failure >> %s",
+        _amqp_get_rpc_reply_string (amqp_r));
+    goto out;
+  }
+  if (!amqp_channel_open (amqp_conn, amqp_channel)) {
+    g_set_error (error, CHAMGE_BACKEND_ERROR,
+        CHAMGE_BACKEND_ERROR_OPERATION_FAILURE, "channel open failure >> %s",
+        _amqp_get_rpc_reply_string (amqp_get_rpc_reply (amqp_conn)));
+    goto out;
+  }
+
+  g_debug ("success to login %s:%d/%s id[%s] pwd[%s]",
+      connection_info.host, connection_info.port,
+      connection_info.vhost, connection_info.user, connection_info.password);
+
+  ret =
+      _amqp_rpc_request (amqp_conn, amqp_channel, cmd,
+      amqp_exchange_name, _get_queue_name (cmd), out, error);
+  if (ret != CHAMGE_RETURN_OK && *error != NULL) {
+    g_debug ("rpc request failure >> %s", (*error)->message);
+    goto out;
+  }
+
+  if (amqp_conn != NULL) {
+    gint destroy_ret = 0;
+    amqp_r = amqp_channel_close (amqp_conn, amqp_channel, AMQP_REPLY_SUCCESS);
+    if (amqp_r.reply_type != AMQP_RESPONSE_NORMAL) {
+      g_set_error (error, CHAMGE_BACKEND_ERROR,
+          CHAMGE_BACKEND_ERROR_OPERATION_FAILURE, "channel close failure >> %s",
+          _amqp_get_rpc_reply_string (amqp_r));
+      ret = CHAMGE_RETURN_FAIL;
+      goto out;
+    }
+    amqp_r = amqp_connection_close (amqp_conn, AMQP_REPLY_SUCCESS);
+    if (amqp_r.reply_type != AMQP_RESPONSE_NORMAL) {
+      g_set_error (error, CHAMGE_BACKEND_ERROR,
+          CHAMGE_BACKEND_ERROR_OPERATION_FAILURE,
+          "connection close failure >> %s",
+          _amqp_get_rpc_reply_string (amqp_r));
+      ret = CHAMGE_RETURN_FAIL;
+      goto out;
+    }
+    destroy_ret = amqp_destroy_connection (amqp_conn);
+    if (destroy_ret < 0) {
+      g_set_error (error, CHAMGE_BACKEND_ERROR,
+          CHAMGE_BACKEND_ERROR_OPERATION_FAILURE,
+          "connection destroy failure >> %s", amqp_error_string2 (destroy_ret));
+      ret = CHAMGE_RETURN_FAIL;
+      goto out;
+    }
+  }
+
+out:
+  return ret;
+}
+
+
+
```

Comment:
nitpick: Separate functions with one empty line. You tend to use 1-3 at random.

Reply:
Sorry, you are right. As probably @justinjoy will require some changes I can fix it in the next version.